File: January 2023/Day 26/Case-specific Sorting of Strings/casespecific-sorting-of-strings.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;


// } Driver Code Ends
//User function Template for Java

class Solution
{
    //Function to perform case-specific sorting of strings.
    // Time Complexity :- O(n.log(n));
    // Space Complexity :- O(n);
    public static String caseSort(String str)
    {
        // Your code here
        char[] ch = str.toCharArray();
        Arrays.sort(ch);
        int up=0,lo=0;
        while(lo<ch.length) {
            if(ch[lo]>='a') break;
            lo++;
        }
        
        char[] ans = new char[ch.length];
        for(int i=0;i<str.length();i++) {
            if(str.charAt(i)>='a') {
                ans[i]=ch[lo++];
            } else {
                ans[i]=ch[up++];
            }
        }
        return new String(ans);
    }
}

//{ Driver Code Starts.

class GFG {
	public static void main (String[] args) {
		
    	try {
    	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	    PrintWriter out=new PrintWriter(System.out);
    	    String[] words = br.readLine().split("\\s+");
    	    int numTestCases = Integer.parseInt(words[0]);
    	    
    	    for (int tIdx = 0; tIdx < numTestCases; tIdx++) {
    	        words = br.readLine().split("\\s+");
    	        int size = Integer.parseInt(words[0]);
    	        String str = br.readLine();
    	        String sortedStr = new Solution().caseSort(str);
    	        out.println(sortedStr);
    	    }
    	    out.close();
    	}
    	catch (IOException e) {
    	    System.out.println(e);
    	}
	}
}
// } Driver Code Ends

File: January 2023/Day 26/Case-specific Sorting of Strings/casespecific-sorting-of-strings.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution
{
    public:
    //Function to perform case-specific sorting of strings.
    string caseSort(string str, int n)
    {
        // your code here
        string s1=str;
        sort(s1.begin(),s1.end());
        int up=0,lo=0;
        while(lo<s1.length()) {
            if(s1[lo]>='a') break;
            lo++;
        }
        
        string ans="";
        for(int i=0;i<str.length();i++) {
            if(str[i]>='a') {
                ans+=s1[lo++];
            } else {
                ans+=s1[up++];
            }
        }
        
        return ans;
    }
};

//{ Driver Code Starts.

int main()
{
	int t;
	cin>>t;
	while(t--)
	{
		int n;
		cin>>n;
		string str;
		cin>>str;
		Solution obj;
		cout<<obj.caseSort (str, n)<<endl;
	}
}
// } Driver Code Ends